lib/set.js

```diff
@@ -2,62 +2,74 @@
 
 const Ref = require('./ref');
 
-module.exports = class Set {
+module.exports = class InternalSet {
 
-    constructor() {
+    constructor(from) {
 
-        this._set = [];
+        this._set = new Set(from);
     }
 
     add(value, refs) {
 
         if (!Ref.isRef(value) && this.has(value, null, null, false)) {
 
-            return;
+            return this;
         }
 
         if (refs !== undefined) { // If it's a merge, we don't have any refs
             Ref.push(refs, value);
         }
 
-        this._set.push(value);
+        this._set.add(value);
         return this;
     }
 
     merge(add, remove) {
 
-        for (let i = 0; i < add._set.length; ++i) {
-            this.add(add._set[i]);
+        for (const item of add._set) {
+            this.add(item);
         }
 
-        for (let i = 0; i < remove._set.length; ++i) {
-            this.remove(remove._set[i]);
+        for (const item of remove._set) {
+            this.remove(item);
         }
 
         return this;
     }
 
     remove(value) {
 
-        this._set = this._set.filter((item) => value !== item);
+        this._set.delete(value);
         return this;
     }
 
     has(value, state, options, insensitive) {
 
-        for (let i = 0; i < this._set.length; ++i) {
-            let items = this._set[i];
+        if (this._set.size === 0) {
+            return false;
+        }
+
+        const isSimpleValue = !(value instanceof Date) && !(insensitive && typeof value === 'string') && !Buffer.isBuffer(value);
```

Comment:
It's simpler for sure, and should work. It might reduce the performance slightly for some cases, though it will speed up others tremendously. Eg. insensitive string lookup, where it actually matches the case.

I expect it will be a net win.